### PR TITLE
Copying MCAD CRDs in Kustomize config instead of referencing them

### LINKED ISSUE #353: Copy MCAD CRDs in Kustomize config instead of referencing them
Currently MCAD CRDs are included by references: https://github.com/project-codeflare/codeflare-operator/blob/1268065a9b2f005f610ad1981ad4b4f5689e1f84/config/crd/mcad/kustomization.yaml#L4

This used to work as they were dereferenced during the build of the OLM bundle.

Now that the CFO is managed by the ODH operator v2 using plain Kustomize manifests, they must be dereferenced beforehand, also because it'll be required for disconnected environments.

That can be added as extra commands run in the `manifests` target for example:

```
$(SED) -i -E "s|(- )\${MCAD_REPO}.*|\1\${MCAD_CRD}|" config/crd/mcad/kustomization.yaml
$(KUSTOMIZE) build config/crd/mcad > config/crd/mcad.yaml
```
 
And remove the inclusion of the `mcad` directory from the `config/crd/kustomization.yaml` file.

### PR DESCRIPTION
**Pull request by Fiona-Waters**
# Issue link
Closes #353 

# What changes have been made
Refactored to copy MCAD CRDs instead of referencing them.

# Verification steps
- Run `make manifests` and ensure that the config/crd/mcad/kustomization.yaml shows the mcad version (not 0.0.0.0).
- Install MCAD in a cluster by running `make install` and ensure that the required CRDs have been created and that everything runs as expected.

## Checks
- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [X] Manual tests
   - [ ] Testing is not required for this change

### CODE REVIEW

File: config/crd/kustomization.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # This kustomization.yaml is not intended to be run by itself,
 # since it depends on service name and namespace that are out of this kustomize package.
 # It should be run by config/default
-resources:
-- mcad
```
Comment (astefanutti) [inline config/crd/kustomization.yaml:5]: Should `mcad.yaml` be added there?

File: Makefile [hunk 2]
```diff
@@ -147,8 +147,10 @@ defaults:
 	gofmt -w $(DEFAULTS_TEST_FILE)
 
 .PHONY: manifests
-manifests: controller-gen ## Generate RBAC objects.
+manifests: controller-gen kustomize ## Generate RBAC objects.
 	$(CONTROLLER_GEN) rbac:roleName=manager-role webhook paths="./..."
+	$(SED) -i -E "s|(- )\${MCAD_REPO}.*|\1\${MCAD_CRD}|" config/crd/mcad/kustomization.yaml
```
Comment (astefanutti) [inline Makefile:152]: Does that make the other `sed` commands from the dependant targets redundant?

File: hack/verify-manifests.sh [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+manifest_file=config/crd/mcad.yaml
+appwrappers=appwrappers.workload.codeflare.dev
+quotasubtrees=quotasubtrees.quota.codeflare.dev
+schedulingspecs=schedulingspecs.workload.codeflare.dev
+
+echo $manifest_file
+if grep $appwrappers $manifest_file \
```
Comment (astefanutti) [inline hack/verify-manifests.sh:9]: We would rather want to check there is no diff, than simply check the files are present.

File: config/crd/kustomization.yaml (hunk 1 above)
Comment (zdtsw) [reply config/crd/kustomization.yaml:5]: i probably will take  a different approach.
Have 3 CRD yaml files in this `config/crd` , not combined all into 1 mcad.yaml
it is easier to compare if upstream has CRD changes and only apply changed yaml.

Comment (astefanutti) [reply config/crd/kustomization.yaml:5]: I agree with @zdtsw that it'd be ideally to have them separated into different files.

File: Makefile [hunk 4]
```diff
@@ -146,9 +146,13 @@ defaults:
...
 	gofmt -w $(DEFAULTS_TEST_FILE)
 
+# this encounters sed issues on MacOS, quick fix is to use gsed or to escape the parentheses i.e. \( \)
 .PHONY: manifests
-manifests: controller-gen ## Generate RBAC objects.
+manifests: controller-gen kustomize ## Generate RBAC objects.
 	$(CONTROLLER_GEN) rbac:roleName=manager-role webhook paths="./..."
+	$(SED) -i -E "s|(- )\${MCAD_REPO}.*|\1\${MCAD_CRD}|" config/crd/mcad/kustomization.yaml
+	$(KUSTOMIZE) build config/crd/mcad > config/crd/mcad.yaml && make split_yaml FILE=config/crd/mcad.yaml
+	git restore config/*
```
Comment (astefanutti) [inline Makefile:155]: I think that should be removed.

**Review by dimakis — COMMENTED** [review]
/lgtm

File: Makefile [hunk 5]
```diff
@@ -388,3 +388,32 @@ verify-imports: openshift-goimports ## Run import verifications.
...
+
+
+FILE ?= input.yaml  # Default value, it isn't a file, but the make cmds fill hang for longer without it
+temp_dir := temp_split
+output_dir := 'config/crd/'
+
+.PHONY: check_yq
+check_yq:
+	@command -v yq >/dev/null 2>&1 || (echo "Installing wget..."; yum install -y wget)
+	@command -v yq >/dev/null 2>&1 || (echo "Installing yq..."; wget https://github.com/mikefarah/yq/releases/download/v4.2.0/yq_linux_amd64.tar.gz -O - |\
```
Comment (dimakis) [inline Makefile:400]: this needs to be YQ

File: Makefile [hunk 6]
```diff
@@ -388,3 +388,32 @@ verify-imports: openshift-goimports ## Run import verifications.
...
 	$(OPERATOR_SDK) scorecard bundle
+
+
+FILE ?= input.yaml  # Default value, it isn't a file, but the make cmds fill hang for longer without it
+temp_dir := temp_split
+output_dir := 'config/crd/'
+
+.PHONY: check_yq
+check_yq:
+	@command -v yq >/dev/null 2>&1 || (echo "Installing wget..."; yum install -y wget)
```
Comment (dimakis) [inline Makefile:399]: this should be 
@command -v wget

File: Makefile [hunk 7]
```diff
@@ -388,3 +388,32 @@ verify-imports: openshift-goimports ## Run import verifications.
...
+	@$(MAKE) check_yq
+	@mkdir -p $(temp_dir)
+	@awk '/apiVersion: /{if (x>0) close("$(temp_dir)/section_" x ".yaml"); x++}{print > "$(temp_dir)/section_"x".yaml"}' $(FILE)
+	@$(MAKE) process_sections
+
+.PHONY: process_sections
+process_sections:
+	@mkdir -p $(output_dir)
+	@for section_file in $(temp_dir)/section_*; do \
+		metadata_name=$$(yq e '.metadata.name' $$section_file); \
```
Comment (dimakis) [inline Makefile:415]: Here you're calling YQ

**Comment by VanillaSpoon** [comment]
If you check out this: https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md

On ubuntu latest wget is preinstalled, and yum is not :)

File: Makefile [hunk 8]
```diff
@@ -146,9 +146,12 @@ defaults:
 
 	gofmt -w $(DEFAULTS_TEST_FILE)
 
+# this encounters sed issues on MacOS, quick fix is to use gsed or to escape the parentheses i.e. \( \)
 .PHONY: manifests
-manifests: controller-gen ## Generate RBAC objects.
+manifests: controller-gen kustomize ## Generate RBAC objects.
 	$(CONTROLLER_GEN) rbac:roleName=manager-role webhook paths="./..."
+	$(SED) -i -E "s|(- )\${MCAD_REPO}.*|\1\${MCAD_CRD}|" config/crd/mcad/kustomization.yaml
+	$(KUSTOMIZE) build config/crd/mcad > config/crd/mcad.yaml && make split_yaml FILE=config/crd/mcad.yaml
```
Comment (dimakis) [inline Makefile:154]: dont call this

File: Makefile [hunk 9]
```diff
@@ -146,9 +146,12 @@ defaults:
 
 	gofmt -w $(DEFAULTS_TEST_FILE)
 
+# this encounters sed issues on MacOS, quick fix is to use gsed or to escape the parentheses i.e. \( \)
 .PHONY: manifests
-manifests: controller-gen ## Generate RBAC objects.
+manifests: controller-gen kustomize ## Generate RBAC objects.
 	$(CONTROLLER_GEN) rbac:roleName=manager-role webhook paths="./..."
+	$(SED) -i -E "s|(- )\${MCAD_REPO}.*|\1\${MCAD_CRD}|" config/crd/mcad/kustomization.yaml
```
Comment (dimakis) [inline Makefile:153]: here too

**Pull request merged**